### PR TITLE
JDK-8328886: Lilliput: Build COH archives

### PR DESCRIPTION
**Pull request by tstuefe**
This is a breakout from and a prerequisite for [JDK-8325104](https://bugs.openjdk.org/browse/JDK-8325104).

To run with +COH and CDS enabled, and to enable tests testing this combination, we need to build archives with +COH too.

This patch expands the number of archives built to four:

a) **classes.jsa** (+UseCompressedClassPointers, -UseCompactObjectHeaders)
b) **classes_nocoops.jsa** (-UseCompressedClassPointers, -UseCompactObjectHeaders)
c) **classes_coh.jsa** (+UseCompressedClassPointers, +UseCompactObjectHeaders)
d) **classes_nocoops_coh.jsa** (-UseCompressedClassPointers, +UseCompactObjectHeaders)

Naming could be more logical, but my aim was to keep compatibility with upstream archive names (no functional or name changes for *classes.jsa* and *classes_nocoops.jsa*). 

Another design goal was for the name-to-options mapping not to change *even if UseCOH were enabled by default*. So, _classes.jsa_ is not the "default jsa" but explicitly and always signifies the jsa with +coops -coh.

The new configure option `--(enable|disable)-cds-archive-coh` allows you to enable/disable the generation of COH archives. By default, it is enabled if the underlying platform supports COH (64-bit) and if standard CDS archive generation is enabled (e.g., not for cross-compilation).

Note that the upcoming JEP "CDS Archived Object Streaming" ([JDK-8326035](https://bugs.openjdk.org/browse/JDK-8326035)) will eliminate the +-Coops distinction, thus reducing the number of CDS archives. It may also eliminate the +-COH distinction at some point, though that is harder.

But that JEP has to be developed and shipped first. After that, it will run alongside the old implementation for a while, which we keep as a fallback. Only if it is the sole remaining way to load CDS archives can we cut down the number of jsa archives. Until then, I propose to keep +-COH archives in addition to +-Coops archives.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8328886](https://bugs.openjdk.org/browse/JDK-8328886): Lilliput: Build COH archives (**Enhancement** - P4)

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/lilliput.git pull/145/head:pull/145` \
`$ git checkout pull/145`

Update a local copy of the PR: \
`$ git checkout pull/145` \
`$ git pull https://git.openjdk.org/lilliput.git pull/145/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 145`

View PR using the GUI difftool: \
`$ git pr show -t 145`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/lilliput/pull/145.diff">https://git.openjdk.org/lilliput/pull/145.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/lilliput/pull/145#issuecomment-2020096896)

### CODE REVIEW

**Comment by tstuefe** [comment]
Linux x64 minimal, Linux aarch64 errors unrelated. Bintools build failed with "fatal error: sys/sdt.h: No such file or directory"

**Comment by tstuefe** [comment]
Windows aarch64 error is mysterious, but from googling around, it looks like a GitHub issue. Unfortunately, I have no Windows aarch64 machine to repeat the test manually.

File: make/Images.gmk [hunk 1]
```diff
@@ -130,10 +130,16 @@ CDS_DUMP_FLAGS = -Xmx128M -Xms128M
...
 #
 # Param1 - VM variant (e.g., server, client, zero, ...)
-# Param2 - _nocoops, or empty
+# Param2 - _nocoops, _coh, _nocoops_coh, or empty
 define CreateCDSArchive
-  $1_$2_DUMP_EXTRA_ARG := $(if $(filter _nocoops, $2),-XX:-UseCompressedOops,)
-  $1_$2_DUMP_TYPE      := $(if $(filter _nocoops, $2),-NOCOOPS,)
+  $1_$2_COOPS_OPTION := $(if $(findstring _nocoops, $2),-XX:-UseCompressedOops)
+  # enable and also explicitly disable coh as needed.
+  ifeq ($(call isTargetCpuBits, 64), true)
```
Comment (rkennke) [inline make/Images.gmk:137]: Not all 64-bit platforms currently support COH. Only aarch64 and x86_64 do. Would that be a problem?

**Review by rkennke — COMMENTED** [review]
Looks good, but I have a question.

File: make/Images.gmk (hunk 1 above)
Comment (tstuefe) [reply make/Images.gmk:137]: Okay, fixed and tested. We now only build COH archives when on x64 or on aarch64, and for the latter, only when not crossbuilding.

Tested on Linux x64 (for x64, x32, aarch64 crossbuild) and Mac m1 (aarch64 native)

**Review by rkennke — APPROVED** [review]
Looks good now, thanks!

**Comment by tstuefe** [comment]
> Looks good now, thanks!

Okay, will wait for GHAs to finish, then push.

**Comment by tstuefe** [comment]
Remaining x86 problems unrelated.

/integrate

**Pull request closed without merging**